core/impl/src/main/java/com/blazebit/persistence/impl/AbstractCTECriteriaBuilder.java

```diff
@@ -201,38 +209,83 @@ public CTEInfo createCTEInfo() {
-            List<Attribute<?, ?>> attributePath = attributeEntry.getAttributePath();
-            attributes.add(attributeName);
-
-            if (JpaMetamodelUtils.isJoinable(attributePath.get(attributePath.size() - 1))) {
-                // We have to map *-to-one relationships to their ids
-                EntityType<?> type = mainQuery.metamodel.entity(attributeEntry.getElementClass());
-                Attribute<?, ?> idAttribute = JpaMetamodelUtils.getSingleIdAttribute(type);
+        final Queue<String> attributeQueue = new ArrayDeque<>(bindingMap.keySet());
+        while (!attributeQueue.isEmpty()) {
+            final String attributeName = attributeQueue.remove();
+            Integer tupleIndex = bindingMap.get(attributeName);
+
+            final ExtendedAttribute attributeEntry = attributeEntries.get(attributeName);
+            final List<Attribute<?, ?>> attributePath = attributeEntry.getAttributePath();
+            final JpaMetamodelAccessor jpaMetamodelAccessor = mainQuery.jpaProvider.getJpaMetamodelAccessor();
+            final Attribute<?, ?> lastAttribute = attributePath.get(attributePath.size() - 1);
+
+            if (jpaMetamodelAccessor.isJoinable(lastAttribute) || lastAttribute.getPersistentAttributeType() == Attribute.PersistentAttributeType.EMBEDDED) {
+                // We have to map *-to-one relationships to their id or unique props
                 // NOTE: Since we are talking about *-to-ones, the expression can only be a path to an object
                 // so it is safe to just append the id to the path
-                Expression selectExpression = selectManager.getSelectInfos().get(bindingEntry.getValue()).getExpression();
+                Expression selectExpression = selectManager.getSelectInfos().get(tupleIndex).getExpression();
 
                 // TODO: Maybe also allow Treat, Case-When, Array?
                 if (selectExpression instanceof NullExpression) {
                     // When binding null, we don't have to adapt anything
                 } else if (selectExpression instanceof PathExpression) {
-                    PathExpression pathExpression = (PathExpression) selectExpression;
-                    // Only append the id if it's not already there
-                    if (!idAttribute.getName().equals(pathExpression.getExpressions().get(pathExpression.getExpressions().size() - 1).toString())) {
-                        pathExpression.getExpressions().add(new PropertyExpression(idAttribute.getName()));
+                    boolean firstBinding = true;
+                    final Collection<String> embeddedPropertyNames;
+
+                    if (lastAttribute.getPersistentAttributeType() == Attribute.PersistentAttributeType.EMBEDDED) {
+                        embeddedPropertyNames = new TreeSet<>(JpaMetamodelUtils.getEmbeddedPropertyNames((EmbeddableType<?>) ((SingularAttribute<?, ?>) lastAttribute).getType()));
```

Comment:
Actually the implementation for Hibernate of `getIdentifierOrUniqueKeyEmbeddedPropertyNames` already is able to extract keys from component types, because nested id types are wrapped into virtual component types and it was mandatory to account for that. Needless to say, I think it is neater here to treat the embedded attribute separately here. Less room for errors across various ORMs.

Do you think the casts are safe here?

Reply:
Yep, seems ok